File: Sources/ClaudeRelaySpeech/WhisperTranscriber.swift
import Foundation
import WhisperKit

/// Protocol for speech transcription — enables mock injection in tests.
public protocol SpeechTranscribing: Sendable {
    func transcribe(_ audioBuffer: [Float], skipNoSpeechFilter: Bool) async throws -> String
}

extension SpeechTranscribing {
    public func transcribe(_ audioBuffer: [Float]) async throws -> String {
        try await transcribe(audioBuffer, skipNoSpeechFilter: false)
    }
}

/// Wraps WhisperKit to transcribe [Float] audio buffers into text.
/// Shared singleton so the model stays loaded across session switches.
///
/// Lives on `@MainActor` because every call site is already main-actor-isolated
/// (`OnDeviceSpeechEngine` and SwiftUI views). This lets us hold mutable
/// `whisperKit` state safely under Swift 6 without the `@unchecked Sendable`
/// escape hatch.
@MainActor
public final class WhisperTranscriber: SpeechTranscribing {

    public static let shared = WhisperTranscriber()

    private var whisperKit: WhisperKit?
    public private(set) var isLoaded = false

    public init() {}

    /// Download and load the Whisper small.en model.
    /// WhisperKit manages its own model storage under Application Support.
    /// - Parameter progressCallback: Reports 0.0–1.0 as WhisperKit transitions
    ///   through loading → loaded → prewarming → prewarmed.
    public func loadModel(progressCallback: (@Sendable (Double) -> Void)? = nil) async throws {
        // Init without loading so we can attach the state callback first.
        let kit = try await WhisperKit(
            model: "openai_whisper-small.en",
            verbose: false,
            prewarm: false,
            load: false
        )

        if let progressCallback {
            kit.modelStateCallback = { _, newState in
                switch newState {
                case .loading:     progressCallback(0.1)
                case .loaded:      progressCallback(0.5)
                case .prewarming:  progressCallback(0.7)
                case .prewarmed:   progressCallback(1.0)
                default: break
                }
            }
        }

        try await kit.loadModels()
        try await kit.prewarmModels()

        // Clear the state callback so it doesn't fire during transcription.
        // WhisperKit re-validates model state internally and emits transitions
        // that would otherwise re-trigger the caller's progress UI.
        kit.modelStateCallback = nil

        self.whisperKit = kit
        self.isLoaded = true
    }

    /// Transcribe a 16kHz mono Float32 audio buffer.
    /// - Parameter skipNoSpeechFilter: When true, disables the no-speech
    ///   threshold (safe when VAD already confirmed speech) but keeps the
    ///   compression-ratio threshold active to catch decoder repetition loops.
    public func transcribe(
        _ audioBuffer: [Float],
        skipNoSpeechFilter: Bool
    ) async throws -> String {
        guard let whisperKit else {
            throw TranscriberError.modelNotLoaded
        }

        var options: DecodingOptions? = nil
        if skipNoSpeechFilter {
            var opts = DecodingOptions()
            opts.noSpeechThreshold = nil
            options = opts
        }

        let results: [TranscriptionResult] = try await whisperKit.transcribe(
            audioArray: audioBuffer,
            decodeOptions: options
        )

        let text = results
            .map { $0.text }
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)

        guard !text.isEmpty else {
            throw TranscriberError.emptyTranscription
        }

        return Self.collapseRepetitions(text)
    }

    // Two tiers: sentence-level catches punctuated loops, char-level catches unpunctuated
    nonisolated static func collapseRepetitions(_ text: String) -> String {
        let len = text.count
        guard len > 20 else { return text }

        if let collapsed = collapseSentenceRepetitions(text) {
            return collapsed
        }

        if let collapsed = collapseCharacterRepetitions(text) {
            return collapsed
        }

        return text
    }

    /// Split on sentence-ending punctuation and check if segments are
    /// near-identical (handles "Check dir? Check dir? Check dir?").
    nonisolated private static func collapseSentenceRepetitions(_ text: String) -> String? {
        let segments = text
            .components(separatedBy: CharacterSet(charactersIn: ".?!"))
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }

        guard segments.count >= 2 else { return nil }

        let reference = segments[0].lowercased()
        guard reference.count >= 8 else { return nil }

        let matchCount = segments.filter { seg in
            let normalized = seg.lowercased()
            if normalized == reference { return true }
            let dist = levenshteinDistance(normalized, reference)
            return dist <= max(2, reference.count / 5)
        }.count

        guard matchCount == segments.count else { return nil }

        let originalPunctuation = text.last.map { CharacterSet(charactersIn: ".?!").contains(Unicode.Scalar(String($0))!) } ?? false
        let suffix = originalPunctuation ? String(text.last!) : ""
        return segments[0] + suffix
    }

    /// Find the shortest repeating prefix unit (handles direct concatenation
    /// like "check dircheck dircheck dir").
    nonisolated private static func collapseCharacterRepetitions(_ text: String) -> String? {
        let len = text.count
        let minUnit = 10
        guard len >= minUnit * 2 else { return nil }

        for unitLen in minUnit...(len / 2) {
            let unit = String(text.prefix(unitLen))
            let count = len / unitLen
            guard count >= 2 else { continue }
            let reconstructed = String(repeating: unit, count: count)
            guard text.hasPrefix(reconstructed) else { continue }
            let remainder = String(text.dropFirst(reconstructed.count))
            let trimmedRemainder = remainder.trimmingCharacters(in: .whitespacesAndNewlines)
            let tolerance = max(3, unit.count / 3)
            if trimmedRemainder.isEmpty || levenshteinDistance(trimmedRemainder, unit) <= tolerance {
                return unit.trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }

        return nil
    }

    nonisolated private static func levenshteinDistance(_ s: String, _ t: String) -> Int {
        let sArr = Array(s)
        let tArr = Array(t)
        let sLen = sArr.count, tLen = tArr.count
        if sLen == 0 { return tLen }
        if tLen == 0 { return sLen }
        var prev = Array(0...tLen)
        var curr = [Int](repeating: 0, count: tLen + 1)
        for i in 1...sLen {
            curr[0] = i
            for j in 1...tLen {
                let cost = sArr[i-1] == tArr[j-1] ? 0 : 1
                curr[j] = min(prev[j] + 1, curr[j-1] + 1, prev[j-1] + cost)
            }
            prev = curr
        }
        return prev[tLen]
    }

    /// Release the model from memory.
    public func unload() async {
        await whisperKit?.unloadModels()
        whisperKit = nil
        isLoaded = false
    }
}

public enum TranscriberError: Error, LocalizedError {
    case modelNotLoaded
    case emptyTranscription

    public var errorDescription: String? {
        switch self {
        case .modelNotLoaded: return "Whisper model not loaded"
        case .emptyTranscription: return "No speech detected"
        }
    }

    /// Phrases Whisper commonly hallucinates from silence or background noise.
    /// Compared case-insensitively after stripping punctuation.
    public static let silenceHallucinations: Set<String> = [
        "thank you",
        "thanks",
        "thanks for watching",
        "thank you for watching",
        "bye",
        "goodbye",
        "subscribe",
        "like and subscribe",
        "see you next time",
        "see you",
        "you",
        "the end",
        "so",
        "okay",
        "hmm",
        "oh",
        "ah"
    ]

    /// Returns true if the text is a known Whisper silence hallucination.
    public static func isSilenceHallucination(_ text: String) -> Bool {
        let normalized = text
            .lowercased()
            .components(separatedBy: CharacterSet.letters.inverted)
            .joined(separator: " ")
            .split(separator: " ")
            .joined(separator: " ")
        return silenceHallucinations.contains(normalized)
    }
}

File: ClaudeRelayAndroid/speech/src/main/kotlin/relay/speech/transcribe/WhisperTranscriber.kt
package relay.speech.transcribe

import kotlin.math.max

/**
 * Transcribes 16 kHz mono Float32 audio into text via whisper.cpp.
 *
 * Faithful port of `Sources/ClaudeRelaySpeech/WhisperTranscriber.swift`. Only the
 * pure heuristics — [collapseRepetitions] and [isSilenceHallucination] — are
 * implemented here. The native inference path is a deferred EXECUTION NOTE (see
 * [transcribe]).
 */
object WhisperTranscriber {

    /**
     * EXECUTION NOTE — whisper.cpp JNI transcription (DEVICE/CMAKE-DEFERRED, NOT
     * implemented here).
     *
     * Port of `WhisperTranscriber.transcribe(_:skipNoSpeechFilter:)`. The Swift
     * version wraps WhisperKit (CoreML / ANE). The Android equivalent loads the
     * **same Whisper small.en weights** but in **ggml format** (`ggml-small.en.bin`,
     * the whisper.cpp binary format, NOT the WhisperKit CoreML package) and calls
     * into a whisper.cpp JNI bridge.
     *
     * Contract the deferred implementation must satisfy:
     *  - Signature: `suspend fun transcribe(audioBuffer: FloatArray, skipNoSpeechFilter: Boolean): String`.
     *  - Runs the native `whisper_full` call on `Dispatchers.Default` (off the
     *    main/UI thread) — the model is a long-lived singleton kept loaded across
     *    session switches, mirroring `WhisperTranscriber.shared`.
     *  - `skipNoSpeechFilter = true` disables whisper's no-speech threshold
     *    (`no_speech_thold`) because VAD already confirmed speech, but keeps the
     *    compression-ratio threshold active to catch decoder repetition loops.
     *  - Joins all segment texts with " ", trims whitespace, throws
     *    [TranscriberError.EmptyTranscription] if the result is blank, and finally
     *    pipes the text through [collapseRepetitions] before returning.
     *  - Throws [TranscriberError.ModelNotLoaded] when called before the model is
     *    loaded.
     *
     * NOT implemented: there is no CMake / NDK native-build toolchain in this
     * environment and no device to run on, so the whisper.cpp JNI bridge is a
     * later milestone. The pure heuristics below are the testable deliverable.
     */
    @Suppress("UnusedParameter")
    suspend fun transcribe(audioBuffer: FloatArray, skipNoSpeechFilter: Boolean): String =
        throw TranscriberError.ModelNotLoaded

    /**
     * Collapse Whisper decoder repetition loops.
     *
     * Two tiers, in order:
     *  1. [collapseSentenceRepetitions] — catches punctuation-delimited loops
     *     ("Check dir? Check dir? Check dir?").
     *  2. [collapseCharacterRepetitions] — catches unpunctuated direct
     *     concatenation ("check dircheck dircheck dir").
     *
     * Gated on `length > 20`: short transcripts are returned verbatim (a real
     * "go go go" must survive).
     */
    fun collapseRepetitions(text: String): String {
        val len = text.length
        if (len <= 20) return text

        collapseSentenceRepetitions(text)?.let { return it }
        collapseCharacterRepetitions(text)?.let { return it }

        return text
    }

    /**
     * Split on sentence-ending punctuation (`.?!`) and collapse if every segment
     * is near-identical to the first (Levenshtein <= `max(2, ref.length / 5)`).
     * Returns null when no collapse applies.
     */
    private fun collapseSentenceRepetitions(text: String): String? {
        val segments = text
            .split(Regex("[.?!]"))
            .map { it.trim() }
            .filter { it.isNotEmpty() }

        if (segments.size < 2) return null

        val reference = segments[0].lowercase()
        if (reference.length < 8) return null

        val matchCount = segments.count { seg ->
            val normalized = seg.lowercase()
            if (normalized == reference) {
                true
            } else {
                levenshteinDistance(normalized, reference) <= max(2, reference.length / 5)
            }
        }

        if (matchCount != segments.size) return null

        val lastChar = text.lastOrNull()
        val originalPunctuation = lastChar != null && lastChar in ".?!"
        val suffix = if (originalPunctuation) lastChar.toString() else ""
        return segments[0] + suffix
    }

    /**
     * Find the shortest repeating prefix unit (>= 10 chars) that reconstructs the
     * whole string, tolerating a near-identical trailing remainder
     * (Levenshtein <= `max(3, unit.length / 3)`). Returns null when no collapse
     * applies.
     */
    private fun collapseCharacterRepetitions(text: String): String? {
        val len = text.length
        val minUnit = 10
        if (len < minUnit * 2) return null

        for (unitLen in minUnit..(len / 2)) {
            val unit = text.substring(0, unitLen)
            val count = len / unitLen
            if (count < 2) continue
            val reconstructed = unit.repeat(count)
            if (!text.startsWith(reconstructed)) continue
            val remainder = text.substring(reconstructed.length)
            val trimmedRemainder = remainder.trim()
            val tolerance = max(3, unit.length / 3)
            if (trimmedRemainder.isEmpty() || levenshteinDistance(trimmedRemainder, unit) <= tolerance) {
                return unit.trim()
            }
        }

        return null
    }

    /** Classic Levenshtein edit distance (two-row DP). */
    private fun levenshteinDistance(s: String, t: String): Int {
        val sLen = s.length
        val tLen = t.length
        if (sLen == 0) return tLen
        if (tLen == 0) return sLen

        var prev = IntArray(tLen + 1) { it }
        var curr = IntArray(tLen + 1)
        for (i in 1..sLen) {
            curr[0] = i
            for (j in 1..tLen) {
                val cost = if (s[i - 1] == t[j - 1]) 0 else 1
                curr[j] = minOf(prev[j] + 1, curr[j - 1] + 1, prev[j - 1] + cost)
            }
            val tmp = prev
            prev = curr
            curr = tmp
        }
        return prev[tLen]
    }

    /**
     * Phrases Whisper commonly hallucinates from silence or background noise.
     * Compared case-insensitively after stripping punctuation. Mirrors
     * `TranscriberError.silenceHallucinations`.
     */
    val silenceHallucinations: Set<String> = setOf(
        "thank you",
        "thanks",
        "thanks for watching",
        "thank you for watching",
        "bye",
        "goodbye",
        "subscribe",
        "like and subscribe",
        "see you next time",
        "see you",
        "you",
        "the end",
        "so",
        "okay",
        "hmm",
        "oh",
        "ah",
    )

    /**
     * Returns true if the text is a known Whisper silence hallucination.
     *
     * Mirrors `TranscriberError.isSilenceHallucination`: lowercase, replace every
     * run of non-letter characters with a single space, trim, then test set
     * membership.
     */
    fun isSilenceHallucination(text: String): Boolean {
        val normalized = text
            .lowercase()
            .split(Regex("[^\\p{L}]+"))
            .filter { it.isNotEmpty() }
            .joinToString(separator = " ")
        return silenceHallucinations.contains(normalized)
    }
}

/** Errors thrown by [WhisperTranscriber]. Mirrors Swift `TranscriberError`. */
sealed class TranscriberError(message: String) : Exception(message) {
    data object ModelNotLoaded : TranscriberError("Whisper model not loaded")
    data object EmptyTranscription : TranscriberError("No speech detected")
}
